threescale/src/proxy.rs

```diff
@@ -92,15 +96,79 @@ pub fn set_app_id_to_cache(user_key: &UserKey, app_id: &AppId) -> Result<(), Cac
 // returns false on set failure
 pub fn set_application_to_cache(key: &str, app: &Application, cas: u32) -> bool {
     info!("setting application with key: {}", key);
-    match set_shared_data(
-        key,
-        Some(&bincode::serialize::<Application>(app).unwrap()),
-        Some(cas),
-    ) {
-        Ok(()) => true,
-        Err(e) => {
-            debug!("set operation failed for key: {} : {:?}", key, e);
-            false
+    let prev_memory_usage = get_cache_pair_size(key) as i32;
+    let serialized_app = bincode::serialize::<Application>(app).unwrap();
+    let memory_delta: i32 = (key.len() as i32) + (serialized_app.len() as i32) - prev_memory_usage;
+    if let Err(e) = set_shared_data(key, Some(&serialized_app), Some(cas)) {
+        debug!("set operation failed for key: {} : {:?}", key, e);
+        return false;
+    }
+    for num_try in 0..3 {
+        match update_shared_memory_size(memory_delta) {
+            Ok(()) => break,
+            Err(e) => debug!("try#{} : failed to update memory counter: {}", num_try, e),
         }
     }
+    true
+}
+
+// returns memory used in bytes for both key and value pair stored.
+fn get_cache_pair_size(key: &str) -> usize {
+    match get_shared_data(key) {
+        Ok((Some(bytes), _)) => key.len() + bytes.len(),
+        Ok((None, _)) => key.len(),
+        Err(_) => 0,
```

Comment:
If this call fails then you'll get back a 0, but the code will happily try to update the key anyway - this needs error handling to bail out early before trying to write incorrect data (which likely won't work if `get_shared_data` fails, but who knows what could be happening in the host side - ie. bugs, cpu-on-fire condition, etc).

Reply:
Added better error handling.